internal/consensus/state.go

```diff
@@ -1458,6 +1448,16 @@ func (cs *State) defaultDoPrevote(height int64, round int32) {
 				return
 			}
 
+			// Validate proposal block, from consensus' perspective
+			err := cs.blockExec.ValidateBlock(cs.state, cs.ProposalBlock)
```

Comment:
@cason if you agree that moving this here aligns the code with the algorithm (so the code stays correct), would it also be worth moving PBTS's `isTimely` check (the whole _if_ in line 1389 above) here?

I know that it wouldn't probably change anything perf-wise, as checking for `isTimely` is cheap, but I now think the code is clearer if all three checks (`ValidateBlock`, `ProcessProposal`, and `isTimely`) are at the same place

Reply:
The corner case that remains is when  `cs.Proposal.POLRound == -1` and `cs.LockedRound == -1` and `cs.ProposalBlock.HashesTo(cs.ValidBlock.Hash())`. This is an optimization for when a block has became valid but we didn't lock it (we receive the 2/3+ prevotes after issuing our precommit).

Actually, I don't even see how this situation can happen. The situation is that a block has received a Polka in a round, then in a future round a propose that _hasn't_ seen this Polka, proposes the same block as a new block (since  `cs.Proposal.POLRound == -1`). This is then the case of two nodes producing the same block "by chance" in different rounds...